scapy/contrib/rpl.py

```diff
@@ -0,0 +1,312 @@
+                   BitField("R", 0, 1),
+                   BitField("reserved1", 0, 5),
+                   IntField("validlifetime", 0xffffffff),
+                   IntField("preflifetime", 0xffffffff),
+                   IntField("reserved2", 0),
+                   IP6Field("prefix", "::1")]
+
+
+class OptTgtDesc(_RPLGuessOption):
+    """
+    Control Option: RPL Target Descriptor
+    """
+    name = "RPL Target Descriptor"
+    fields_desc = [ByteEnumField("otype", 9, RPLOPTSSTR),
+                   ByteField("len", 4),
+                   IntField("descriptor", 0)]
+
+
+class Pad1(_RPLGuessOption):
+    """
+    Control Option: Pad 1 byte
+    """
+    name = "Pad1"
+    fields_desc = [ByteEnumField("otype", 0x00, RPLOPTSSTR)]
+
+
+class PadN(_RPLGuessOption):
+    """
+    Control Option: Pad N bytes
+    """
+    name = "PadN"
+    fields_desc = [ByteEnumField("otype", 0x01, RPLOPTSSTR),
+                   FieldLenField("optlen", None, length_of="optdata", fmt="B"),
+                   StrLenField("optdata", "",
+                               length_from=lambda pkt: pkt.optlen)]
+
+
+# RPL Control Message Handling
+
+
+class DIS(_RPLGuessOption, Packet):
+    """
+    Control Message: DODAG Information Solicitation (DIS)
+    """
+    name = "DODAG Information Solicitation"
+    fields_desc = [ByteField("flags", 0),
+                   ByteField("reserved", 0)]
+
+
+class DIO(_RPLGuessOption, Packet):
+    """
+    Control Message: DODAG Information Object (DIO)
+    """
+    name = "DODAG Information Object"
+    fields_desc = [ByteField("RPLInstanceID", 50),
+                   ByteField("ver", 0),
+                   ShortField("rank", 1),
+                   BitField("G", 1, 1),
+                   BitField("unused1", 0, 1),
+                   BitEnumField("mop", 1, 3, RPLMOP),
+                   BitField("prf", 0, 3),
+                   ByteField("dtsn", 240),
+                   ByteField("flags", 0),
+                   ByteField("reserved", 0),
+                   IP6Field("dodagid", "::1")]
+
+
+class _OptDODAGIDField(IP6Field):
+    """
+    Handle Optional DODAG ID field in DAO
+    """
+    def getfield(self, pkt, s):
+        if pkt.D == 0:
+            return s, None
+        return s[16:], self.m2i(pkt, s[:16])
+
+    def addfield(self, pkt, s, val):
+        if pkt.D == 1:
+            return s + self.i2m(pkt, val)
+        if val:
+            print("RPL DAO 'D' flag is not set but dodagid is given.")
+        return s
+
+
+class DAO(_RPLGuessOption, Packet):
+    """
+    Control Message: Destination Advertisement Object (DAO)
+    """
+    name = "Destination Advertisement Object"
+    fields_desc = [ByteField("RPLInstanceID", 50),
+                   BitField("K", 0, 1),
+                   BitField("D", 0, 1),
+                   BitField("flags", 0, 6),
+                   ByteField("reserved", 0),
+                   ByteField("daoseq", 1),
+                   _OptDODAGIDField("dodagid", None)]
```

Comment:
Can't you use a `ConditionalField` here instead ? The warning isn't that useful

Reply:
Yep, `ConditionalField` makes perfect sense.  Replaced.